code/controllers/subsystem/html_audio.dm

```diff
@@ -0,0 +1,274 @@
+
+/datum/controller/subsystem/html_audio/proc/update_listener_volume(client/listener)
+	for(var/i in 1 to max_channels)
+		var/volume_to_use = 0
+		var/distance = get_dist(channel_assignment[i], listener.mob)
+		var/turf/speaker_turf = get_turf(channel_assignment[i])
+		var/turf/listener_turf = get_turf(listener.mob)
+		if(!channel_assignment[i] || distance >= 10 || (channel_requires_LOS_at_start[i] && !(listener in channel_requires_LOS_at_start_listeners[i])) || (channel_tts_status[i] && listener.prefs.read_preference(/datum/preference/toggle/sound_tts_use_byond_audio)))
+			volume_to_use = 0
+		else
+			if(channel_tts_status[i])
+				volume_to_use = min((1-(1/10*distance))**2, 1) * listener.prefs.read_preference(/datum/preference/numeric/sound_tts_volume) / 100
+			else
+				volume_to_use = min((1-(1/10*distance))**2, 1)
+				//Atmosphere affects sound
+				var/pressure_factor = 1
+				var/datum/gas_mixture/hearer_env = listener_turf.return_air()
+				var/datum/gas_mixture/source_env = speaker_turf.return_air()
+
+				if(hearer_env && source_env)
+					var/pressure = min(hearer_env.return_pressure(), source_env.return_pressure())
+					if(pressure < ONE_ATMOSPHERE)
+						pressure_factor = max((pressure - SOUND_MINIMUM_PRESSURE)/(ONE_ATMOSPHERE - SOUND_MINIMUM_PRESSURE), 0)
+				else //space
+					pressure_factor = 0
+
+				if(distance <= 1)
+					pressure_factor = max(pressure_factor, 0.15) //touching the source of the sound
+
+				volume_to_use *= pressure_factor
+				volume_to_use = clamp(volume_to_use, 0, 1)
+		listener << output(list2params(list(num2text(volume_to_use), "channel_[i]")), "html_audio_player:setVolume")
+
+/datum/controller/subsystem/html_audio/proc/jank_ass_browse_check(checked_person)
+	if (!winexists(checked_person, "html_audio_player"))
+		checked_person << browse(browse_txt, "window=html_audio_player&file=html_audio_player.htm")
+	if (!winexists(checked_person, "html_audio_preview_player"))
+		checked_person << browse(browse_txt, "window=html_audio_preview_player&file=html_audio_preview_player.htm")
+
+/datum/controller/subsystem/html_audio/proc/register_player(atom/movable/player, requires_LOS = FALSE)
+	var/assigned = FALSE
+	for(var/i in 1 to max_channels) // Go through the channels, find the first empty slot and get in there.
+		if(isnull(channel_assignment[i]))
+			assigned = TRUE
+			channel_assignment[i] = player
+			if(requires_LOS)
+				channel_requires_LOS_at_start[i] = TRUE
+			RegisterSignal(player, COMSIG_MOVABLE_MOVED, PROC_REF(handle_player_move))
+			RegisterSignal(player, COMSIG_PARENT_QDELETING, PROC_REF(deregister_player_qdel))
+			break
+	if(!assigned)
+		CRASH("YO WE OUT OF FUCKING CHANNELS ABORT ABORT ABORT TOO MANY FUCKING THINGS GOT HTML AUDIO REEL THAT SHIT IN BRO BACK THAT ASS UP")
+	list_clear_nulls(listeners) // clients be like *poof* mid proc
+	for(var/client/listener in listeners)
+		if(!listener)
+			continue
```

Comment:
You're already performing a null check in the loop iteration itself--`update_listener_volume` itself needs to be client-deletion-resistant.